File: tests/roundtrip/bidirectional/consumer/SuspendUnitSlotTests.cs
using System;
using System.Reflection;
using System.Threading.Tasks;
using NUnit.Framework;
using roundtrip.suspendunitslot;

public class SuspendUnitSlotTests
{
    private static async Task Complete(Func<UnitGate, Task<kotlin.Unit>> start)
    {
        var gate = new UnitGate();
        var task = start(gate);
        Assert.That(task.IsCompleted, Is.False);
        Assert.That(gate.entries, Is.EqualTo(1));
        gate.release();
        Assert.That(await task, Is.SameAs(kotlin.Unit.INSTANCE));
        Assert.That(gate.entries, Is.EqualTo(1));
    }

    [Test]
    public async Task PlainAndGenericUnitSlotsKeepTheirDistinctTaskSignatures()
    {
        var source = new ImmediateUnitSlot();
        Task ordinary = source.read();
        Task plainSlot = ((PlainUnitSlot)source).read();
        Task<kotlin.Unit> generic = ((UnitSlot<kotlin.Unit>)source).read();
        await ordinary;
        await plainSlot;
        Assert.That(await generic, Is.SameAs(kotlin.Unit.INSTANCE));
        Assert.That(ordinary.IsCompletedSuccessfully, Is.True);
        Assert.That(typeof(ImmediateUnitSlot).GetMethod("read",
            BindingFlags.Public | BindingFlags.Instance | BindingFlags.DeclaredOnly).ReturnType,
            Is.EqualTo(typeof(Task)));
        var map = typeof(ImmediateUnitSlot).GetInterfaceMap(typeof(UnitSlot<kotlin.Unit>));
        var index = Array.FindIndex(map.InterfaceMethods, method => method.Name == "read");
        Assert.That(index, Is.GreaterThanOrEqualTo(0));
        Assert.That(map.TargetMethods[index].IsPrivate, Is.True);
        Assert.That(map.TargetMethods[index].ReturnType, Is.EqualTo(typeof(Task<kotlin.Unit>)));
        Assert.That(await SuspendUnitSlotKt.observeLocal(source), Is.SameAs(kotlin.Unit.INSTANCE));
    }

    [Test]
    public async Task GenericInterfaceBaseAndMethodSlotsResumeWithUnit()
    {
        await Complete(gate => ((UnitSlot<kotlin.Unit>)new DelayedUnitSlot(gate)).read());
        await Complete(gate => ((UnitBaseSlot<kotlin.Unit>)new DelayedBaseUnitSlot(gate)).read());
        await Complete(gate => ((MethodUnitSlot<kotlin.Unit>)new GenericMethodUnitSlot(gate)).read("method"));
        var gate = new UnitGate();
        var source = new FurtherUnitSlot(gate);
        var task = ((UnitSlot<kotlin.Unit>)source).read();
        Assert.That(task.IsCompleted, Is.False);
        Assert.That(source.calls, Is.EqualTo(1));
        gate.release();
        Assert.That(await task, Is.SameAs(kotlin.Unit.INSTANCE));
        Assert.That(source.calls, Is.EqualTo(1));
    }

    [Test]
    public async Task ErasedArgumentsAndErasedResultsUseOneCorrectAdapter()
    {
        foreach (var value in new[] { null, kotlin.Unit.INSTANCE })
        {
            var gate = new UnitGate();
            var source = new ErasedParameterUnitSlot(gate);
            var task = ((ErasedUnitSlot<kotlin.Unit>)source).read(value);
            Assert.That(task.IsCompleted, Is.False);
            Assert.That(source.wasNull, Is.EqualTo(value is null));
            gate.release();
            Assert.That(await task, Is.SameAs(kotlin.Unit.INSTANCE));
        }
        var resultGate = new UnitGate();
        var result = ((NullableResultSlot<kotlin.Unit>)new NullableErasedUnitSlot(resultGate)).read();
        Assert.That(result.IsCompleted, Is.False);
        resultGate.release();
        Assert.That(await result, Is.SameAs(kotlin.Unit.INSTANCE));
    }

    [Test]
    public async Task InheritedAbstractAndDefaultBodiesFillGenericSlots()
    {
        Assert.That(await ((UnitSlot<kotlin.Unit>)new InheritedUnitBody()).read(), Is.SameAs(kotlin.Unit.INSTANCE));
        Assert.That(await ((UnitSlot<kotlin.Unit>)new ConcreteUnitBody()).read(), Is.SameAs(kotlin.Unit.INSTANCE));
        var before = SuspendUnitSlotKt.defaultCount();
        Assert.That(await ((UnitSlot<kotlin.Unit>)new DefaultUnitBody()).read(), Is.SameAs(kotlin.Unit.INSTANCE));
        Assert.That(SuspendUnitSlotKt.defaultCount(), Is.EqualTo(before + 1));
    }

    [Test]
    public void FaultsAndCancellationRemainTaskFailuresOnBothPublicFaces()
    {
        foreach (var generic in new[] { false, true })
        {
            var gate = new UnitGate();
            var source = new DelayedUnitSlot(gate);
            Task failure = generic ? ((UnitSlot<kotlin.Unit>)source).read() : source.read();
            Assert.That(failure.IsCompleted, Is.False);
            gate.fail();
            Assert.ThrowsAsync<InvalidOperationException>(async () => { await failure; });
            Assert.That(failure.IsFaulted, Is.True);
            Task cancellation = generic ? ((UnitSlot<kotlin.Unit>)source).read() : source.read();
            Assert.That(cancellation.IsCompleted, Is.False);
            gate.cancel();
            Assert.CatchAsync<OperationCanceledException>(async () => { await cancellation; });
            Assert.That(cancellation.IsCanceled, Is.True);
        }
    }
}

File: tests/roundtrip/consumer/SuspendUnitSlotTests.kt
package roundtriptests.suspendunitslot

import NUnit.Framework.TestAttribute
import NUnit.Framework.Legacy.ClassicAssert.AreEqual as assertEquals
import NUnit.Framework.Legacy.ClassicAssert.IsTrue as assertTrue
import kotlin.coroutines.*
import roundtrip.suspendunitslot.*

private class Completion : Continuation<Any?> {
    var completed = false
    var value: Any? = "pending"
    var failure: Throwable? = null
    override val context: CoroutineContext get() = EmptyCoroutineContext
    override fun resumeWith(result: Result<Any?>) {
        failure = result.exceptionOrNull()
        if (failure == null) value = result.getOrThrow()
        completed = true
    }
}
private fun start(block: suspend () -> Any?): Completion {
    val completion = Completion()
    block.startCoroutine(completion)
    return completion
}
private fun assertUnit(completion: Completion) {
    assertTrue(completion.completed)
    assertTrue(completion.failure == null)
    assertTrue(completion.value === Unit)
}
private fun complete(gate: UnitGate, completion: Completion) {
    assertTrue(!completion.completed)
    assertEquals(1, gate.entries)
    gate.release()
    assertUnit(completion)
    assertEquals(1, gate.entries)
}
private suspend fun observe(slot: UnitSlot<Unit>): Any? = slot.read()
private suspend fun observeBase(slot: UnitBaseSlot<Unit>): Any? = slot.read()
private suspend fun observeErased(slot: ErasedUnitSlot<Unit>, value: Unit?): Any? = slot.read(value)
private suspend fun observeNullable(slot: NullableResultSlot<Unit>): Any? = slot.read()
private suspend fun observeMethod(slot: MethodUnitSlot<Unit>): Any? = slot.read("method frame")

private class ImportedInterfaceBody(private val gate: UnitGate) : UnitSlot<Unit>, PlainUnitSlot {
    override suspend fun read() { gate.pause() }
}
private class ImportedBaseBody(private val gate: UnitGate) : UnitBaseSlot<Unit>() {
    override suspend fun read() { gate.pause() }
}
private class ImportedErasedBody(private val gate: UnitGate) : ErasedUnitSlot<Unit> {
    override suspend fun read(value: Unit?) { gate.pause() }
}
private class ImportedFurtherBody(private val gate: UnitGate) : DelayedUnitSlot(gate) {
    var calls = 0
    override suspend fun read() { calls++; gate.pause() }
}

class SuspendUnitSlotTests {
    @TestAttribute
    fun directDefaultInheritedAndAbstractImplementationsReturnUnit() {
        assertUnit(start { observe(ImmediateUnitSlot()) })
        assertUnit(start { observeLocal(ImmediateUnitSlot()) })
        assertUnit(start { observe(InheritedUnitBody()) })
        assertUnit(start { observe(ConcreteUnitBody()) })
        val before = defaultCount()
        assertUnit(start { observe(DefaultUnitBody()) })
        assertEquals(before + 1, defaultCount())
    }

    @TestAttribute
    fun localAndImportedCallsActuallySuspendAndResume() {
        val interfaceGate = UnitGate()
        complete(interfaceGate, start { observe(DelayedUnitSlot(interfaceGate)) })
        val localGate = UnitGate()
        complete(localGate, start { observeLocal(DelayedUnitSlot(localGate)) })
        val baseGate = UnitGate()
        complete(baseGate, start { observeBase(DelayedBaseUnitSlot(baseGate)) })
        val localBaseGate = UnitGate()
        complete(localBaseGate, start { observeLocalBase(DelayedBaseUnitSlot(localBaseGate)) })
        val methodGate = UnitGate()
        complete(methodGate, start { observeMethod(GenericMethodUnitSlot(methodGate)) })
        val nullableGate = UnitGate()
        complete(nullableGate, start { observeNullable(NullableErasedUnitSlot(nullableGate)) })
        val erasedGate = UnitGate()
        val erased = ErasedParameterUnitSlot(erasedGate)
        complete(erasedGate, start { observeErased(erased, null) })
        assertTrue(erased.wasNull)
    }

    @TestAttribute
    fun separatelyCompiledImplementationsFillReferencedSlots() {
        val interfaceGate = UnitGate()
        complete(interfaceGate, start { observe(ImportedInterfaceBody(interfaceGate)) })
        val baseGate = UnitGate()
        complete(baseGate, start { observeBase(ImportedBaseBody(baseGate)) })
        val erasedGate = UnitGate()
        complete(erasedGate, start { observeErased(ImportedErasedBody(erasedGate), null) })
        val furtherGate = UnitGate()
        val further = ImportedFurtherBody(furtherGate)
        complete(furtherGate, start { observe(further) })
        assertEquals(1, further.calls)
    }

    @TestAttribute
    fun resumedFailuresAndCancellationStayFailures() {
        val failureGate = UnitGate()
        val failure = start { observe(DelayedUnitSlot(failureGate)) }
        assertTrue(!failure.completed)
        failureGate.fail()
        assertTrue(failure.completed)
        assertEquals("unit slot failure", failure.failure!!.message)
        val cancelGate = UnitGate()
        val cancellation = start { observe(ImportedInterfaceBody(cancelGate)) }
        assertTrue(!cancellation.completed)
        cancelGate.cancel()
        assertTrue(cancellation.completed)
        assertEquals("unit slot cancellation", cancellation.failure!!.message)
    }
}
